### DISCUSSION TITLE
integration of ODE (cum AUC) and prediction table

### DISCUSSION THREAD

**Discussion by MahmoudAliAfifi** [nlmixr2/nlmixr2#77] [Q&A]
Dear colleagues, 

I used nlmixr2 to fit simulated data with parallelization, but I faced some problems:
1- values of zeros in the compartment column [A, A2, A3, A4]
2- in A4 I am trying to calculate the  AUC.
3- I do not know how to get the table of predictions.

I attached the code. below

```

do_nlmixrODE_nlme <- function(i) {
  datr<- fread(paste0(".//bootdata//",
                      paste0("bootdata",i,model,"N=", Nsubjects,",n=",ifelse(!sparse,length(time_points),4),
                             ",Frel=", frel_form["Frel"]),".csv"))
  compartmental.infusion.model <- function() {
    ini({
      theta_v1 <- 1.187843 #the log value
      theta_v2	<- 0.6365768  #the log value
      theta_v3<- 	0.6729445  #the log value 
      Q1	<- -5.839602  #the log value
      Q2	<- -3.137296  #the log value
      theta_Vmax<- 	5.181784  #the log value
      theta_Km	<- 6.842683  #the log value
      theta_Ke<- 	-6.119298 #the log value
      # theta_formulation <- 0.04879016 
      eta_v1 ~	0.0217  #the log value
      eta_km	~ 0.121  #the log value
      eta_Ke	~ 0.0292 #the log value
      prop.sd <-  0.00000001  #the log value
      
      
    })
    # and a model block with the error sppecification and model specification
    model({
      Kel<- exp(theta_Ke + eta_Ke) #changed to the exp(log format)
      km<- exp(theta_Km + eta_km)  #changed to the exp(log format)
      V1<-  exp(theta_v1 + eta_v1) #changed to the exp(log format)
      CL<- Kel * V1 #changed to the exp(log format)
      V2<- exp(theta_v2) #changed to the exp(log format)
      V3<- exp(theta_v3)#changed to the exp(log format)
      Q1 <-  exp(Q1)
      Q2 <- exp(Q2)
      C1<- A1/V1
      C2<- A2/V2
      C3<- A3/V3
      CUMAUC= A4
      d/dt(A1)= - Kel*A1 - theta_Vmax * C1/(km + C1) - Q1*(C1 - C2) - Q2*(C1 - C3)
      d/dt(A2)= Q1 * (C1 - C2)
      d/dt(A3)=Q2 * (C1 - C3)
      d/dt(A4)<- C1
      C1 ~ add(prop.sd) # I added a small proportional error.
    })    
    
    
  }
  fit <-
    nlmixr2(
      compartmental.infusion.model,
      data = datr,
      est = "saem")
  save(fit, file = paste("fit_NLME_ODE_UUI_", i, ".Rdata", sep = "")) #the fit will be saved as R object in the directory
}
#install.packages("doParallel")
library(doParallel)
cl <- makeCluster(numCores)
registerDoParallel(cl)
tic()

nlmixr_out <-
  foreach(i = 1:NBOOT, .packages = c('nlmixr2', "data.table")) %dopar% do_nlmixrODE_nlme(i)



```

**Comment by mattfidler** [comment]
You can probably use `bootstapFit()` for this problem.

What exactly do you need from it?

**Comment by MahmoudAliAfifi** [comment]
The scenario that I am trying to do is:

1- From the simulated data ( 1000 subjects/ formulation), I will take samples with replacement (with stratification based on the formulation). 
2- Make 500 bootstrap datasets ( each dataset contains 50/formulation for example)
3- fitting these 500, we need to calculate the Cmax and AUC.
So we need a table that contains C [ to calculate the Cmax at end of infusion] and Auc for each patient ( table like the attached table)
![image](https://user-images.githubusercontent.com/115409927/201091568-0d46aed6-5450-4c0b-ba37-572a90a274f6.png)

repeat steps 2 and 3 for different scenarios ( different sample sizes, sampling, and formulations)

**Comment by MahmoudAliAfifi** [comment]
Dear Matt, do you see a problem in the nlmixr code that could cause all compartments to be equal to zero?

**Comment by mattfidler** [comment]
> Dear Matt, do you see a problem in the nlmixr code that could cause all compartments to be equal to zero?

I don't see an issue in the ODEs; perhaps it is because of the very small proportional error.

Another thing that can make parallel running of nlmixr2/rxode2 more stable is to add a cache directory.  This allows locking and unlocking of models so they don't build on top of one another.  I had forgotten this when you asked initially.

**Comment by MahmoudAliAfifi** [comment]
Thanks a lot Matt,

how to create cache directory , do you mean to add rxCreateCache before ODE?
Just one question, u missed. is there is a method to generate a table with the AUC and predicted Concentration? or I should take it from fit object

**Comment by mattfidler** [comment]
Hi @MahmoudAliAfifi 

You only need to call `rxCreateCache()` once (but then you may have to clean it from time to time with `rxClean()`)

I was trying to get a reproducible example so I can show you how to use `bootstrapFit()` to get the information you want, my computer froze (I think my computer is dying since it happens randomly now).

I am rerunning it now and hope to post it soon.

**Comment by mattfidler** [comment]
Reprex is not working on my linux comupter with Wayland, but here is the code:

```r

library(nlmixr2)
library(tidyverse)

compartmental.infusion.model <- function() {
  ini({
    theta_v1 <- 1.187843 #the log value
    theta_v2	<- 0.6365768  #the log value
    theta_v3<- 	0.6729445  #the log value 
    Q1	<- -5.839602  #the log value
    Q2	<- -3.137296  #the log value
    theta_Vmax<- 	5.181784  #the log value
    theta_Km	<- 6.842683  #the log value
    theta_Ke<- 	-6.119298 #the log value
    # theta_formulation <- 0.04879016 
    eta_v1 ~	0.0217  #the log value
    eta_km	~ 0.121  #the log value
    eta_Ke	~ 0.0292 #the log value
    lnorm.sd <-  0.01  #the log value
  })
  # and a model block with the error sppecification and model specification
  model({
    Kel<- exp(theta_Ke + eta_Ke) #changed to the exp(log format)
    km<- exp(theta_Km + eta_km)  #changed to the exp(log format)
    V1<-  exp(theta_v1 + eta_v1) #changed to the exp(log format)
    CL<- Kel * V1 #changed to the exp(log format)
    V2<- exp(theta_v2) #changed to the exp(log format)
    V3<- exp(theta_v3)#changed to the exp(log format)
    Q1 <-  exp(Q1)
    Q2 <- exp(Q2)
    C1<- A1/V1
    C2<- A2/V2
    C3<- A3/V3
    CUMAUC= A4
    d/dt(A1)= - Kel*A1 - theta_Vmax * C1/(km + C1) - Q1*(C1 - C2) - Q2*(C1 - C3)
    d/dt(A2)= Q1 * (C1 - C2)
    d/dt(A3)=Q2 * (C1 - C3)
    d/dt(A4)<- C1
    C1 ~ lnorm(lnorm.sd) 
  })    
}

# First lets simulate a dataset based on the model above
set.seed(42)

d <- et(amt=1000, dur=3) %>%
  et(seq(0.1, 24, by=0.1)) %>%
  et(id=1:24)

dat <- nlmixr2(compartmental.infusion.model, d, "rxSolve",
               control=rxControl(addDosing=TRUE, returnType="tibble")) %>%
  dplyr::rename(DV=sim) %>%
  dplyr::select(id, evid, cmt, amt, time, DV)


fit <-
  nlmixr2(
    compartmental.infusion.model,
    data = dat,
    est = "saem",
    control=saemControl(print=0))


   tmp <- bootstrapFit(fit, nboot=10)

# The fit is the name of the model:
dir <- list.files(pattern="nlmixr2BootstrapCache_fit")

full <- do.call("rbind",
                lapply(list.files(dir, pattern="fitEns"), function(file) {
                  f <- readRDS(file.path(dir, file))
                  f <- addTable(f)
                  .ret <- suppressWarnings(as.data.frame(f))
                  .ret$replicate <- as.numeric(gsub("[^0-9]*([0-9]+)[^0-9]*", "\\1", file))
                  row.names(.ret) <- NULL
                  .ret
                })) %>%
  as_tibble(.)
```

At the end it will give you something like:

```r
full
# A tibble: 57,600 × 28
   ID     TIME    DV  PRED   RES IPRED   IRES IWRES eta_v1 eta_km   eta_Ke    C1
   <fct> <dbl> <dbl> <dbl> <dbl> <dbl>  <dbl> <dbl>  <dbl>  <dbl>    <dbl> <dbl>
 1 1       0.1  6.23  212. -206.  137. -131.  -8.09  0.437 0.0156 -0.00226  137.
 2 1       0.2 12.4   212. -200.  137. -125.  -6.28  0.437 0.0156 -0.00226  137.
 3 1       0.3 19.0   212. -193.  137. -118.  -5.18  0.437 0.0156 -0.00226  137.
 4 1       0.4 25.3   213. -187.  137. -112.  -4.43  0.437 0.0156 -0.00226  137.
 5 1       0.5 31.2   213. -182.  138. -106.  -3.89  0.437 0.0156 -0.00226  138.
 6 1       0.6 37.5   213. -176.  138. -100.  -3.40  0.437 0.0156 -0.00226  138.
 7 1       0.7 43.9   213. -169.  138.  -93.9 -2.99  0.437 0.0156 -0.00226  138.
 8 1       0.8 50.1   213. -163.  138.  -87.8 -2.65  0.437 0.0156 -0.00226  138.
 9 1       0.9 56.0   214. -158.  138.  -82.0 -2.36  0.437 0.0156 -0.00226  138.
10 1       1   61.9   214. -152.  138.  -76.2 -2.10  0.437 0.0156 -0.00226  138.
# … with 57,590 more rows, and 16 more variables: A1 <dbl>, A2 <dbl>, A3 <dbl>,
#   A4 <dbl>, Kel <dbl>, km <dbl>, V1 <dbl>, CL <dbl>, V2 <dbl>, V3 <dbl>,
#   C2 <dbl>, C3 <dbl>, CUMAUC <dbl>, tad <dbl>, dosenum <dbl>, replicate <dbl>
# ℹ Use `print(n = ...)` to see more rows, and `colnames()` to see all variable names
```

**Comment by mattfidler** [comment]
Probably should transfer this to a discussion.

**Comment by MahmoudAliAfifi** [comment]
Thanks a lot Matt, I appreciate this huge effort. I like the bootstrapfit, I used it before.
but because we will work on grid with hundreds cores we want to make parallelization. bootstrapfit only does not use the all cores, I do not know but my CPU is a around 60%

Is there is a way to use foreach with it or future_map with bootstrapfit or any technique for parallelization?

**Reply by mattfidler** [reply]
Well you could, but it isn't integrated yet into `nlmixr2est`.  You could also simply increase your core usage with `setRxThreads(percent=100)` to see if you are getting CPU usage at almost `100%`.  By default the allocation is about half of the number of threads.

**Comment by MahmoudAliAfifi** [comment]
I have a question Matt on the " full " dataset you printed above.
What should I do to make the "amt" column appears in this dataset?
Because I have several formulations ( the difference between them in the dose) so I can compare the AUC, Cmax between them.

**Reply by mattfidler** [reply]
`rxControl(addDosing=TRUE)`

Gives the dose information.  I believe that you could probably do the same by `saemConrtol(rxControl(addDosing=TRUE))`, but I haven't tested it
